development/tools/pkg/diskscollector/collector.go

```diff
@@ -0,0 +1,137 @@
+package diskscollector
+
+import (
+	"log"
+	"regexp"
+	"time"
+
+	compute "google.golang.org/api/compute/v1"
+)
+
+//go:generate mockery -name=ZoneAPI -output=automock -outpkg=automock -case=underscore
+
+// ZoneAPI allows to acces Zones Compute API in GCP
+type ZoneAPI interface {
+	ListZones(project string) ([]string, error)
+}
+
+//go:generate mockery -name=DiskAPI -output=automock -outpkg=automock -case=underscore
+
+// DiskAPI allows to access Disks Compute API in GCP
+type DiskAPI interface {
+	ListDisks(project, zone string) ([]*compute.Disk, error)
+	RemoveDisk(name, project, zone string) error
+}
+
+// DisksGarbageCollector can find and delete disks provisioned by gke-integration prow jobs and not cleaned up properly.
+type DisksGarbageCollector struct {
+	zoneAPI      ZoneAPI
+	diskAPI      DiskAPI
+	shouldRemove DiskRemovalPredicate
+}
+
+// NewDisksGarbageCollector returns a new instance of DisksGarbageCollector
+func NewDisksGarbageCollector(zoneAPI ZoneAPI, diskAPI DiskAPI, shouldRemove DiskRemovalPredicate) *DisksGarbageCollector {
+	return &DisksGarbageCollector{zoneAPI, diskAPI, shouldRemove}
+}
+
+// Run executes disks garbage collection process
+func (gc *DisksGarbageCollector) Run(project string, makeChanges bool) error {
+
+	garbageDisks, err := gc.list(project)
+	if err != nil {
+		return err
+	}
+
+	for _, gd := range garbageDisks {
+
+		var err error
+		var msgPrefix string
+
+		if makeChanges {
+			err = gc.diskAPI.RemoveDisk(project, gd.zone, gd.disk.Name)
+		} else {
+			msgPrefix = "[DRY RUN] "
+		}
+
+		if err != nil {
+			log.Printf("Error deleting disk %s: %#v", gd.disk.Name, err)
+		} else {
+			log.Printf("%sRequested disk delete: \"%s\". Project \"%s\", zone \"%s\", disk creationTimestamp: \"%s\"", msgPrefix, gd.disk.Name, project, gd.zone, gd.disk.CreationTimestamp)
+		}
+	}
+
+	return nil
+}
+
+type garbageDisk struct {
+	zone string
+	disk *compute.Disk
+}
+
+// list returns a filtered list of all disks in the project
+// The list contains only disks that match removal criteria
+func (gc *DisksGarbageCollector) list(project string) ([]*garbageDisk, error) {
+	zones, err := gc.zoneAPI.ListZones(project)
+
+	if err != nil {
+		return nil, err
+	}
+
+	toRemove := []*garbageDisk{}
+
+	for _, zone := range zones {
+		disks, err := gc.diskAPI.ListDisks(project, zone)
+		if err != nil {
+			log.Printf("Error listing disks for zone \"%s\": %#v", zone, err)
```

Comment:
No. This tool does "best effort" job in case of failures. There are multiple zones, temporary failure in one of them doesn't mean we shouldn't continue with another. This will be used periodically: The fact that we didn't manage to clean ALL disks in one run should be logged, but we'll repeat it anyway.